meson.build

```diff
@@ -514,73 +507,52 @@ if not posix and not win32_desktop
 endif
 
 glob_posix = cc.has_function('glob', prefix: '#include <glob.h>')
-if glob_posix
-    features += 'glob_posix'
-endif
+features += {'glob-posix': glob_posix}
 
 glob_win32 = win32 and not posix
 if glob_win32
-    features += 'glob_win32'
     sources += files('osdep/glob-win.c')
 endif
+features += {'glob-win32': glob_win32}
 
 glob = glob_posix or glob_win32
-if glob
-    features += 'glob'
-endif
+features += {'glob': glob}
 
 vt_h = cc.has_header_symbol('sys/vt.h', 'VT_GETMODE')
-if vt_h
-    features += 'vt.h'
-endif
+features += {'vt.h': vt_h}
 
 consio_h = not vt_h and cc.has_header_symbol('sys/consio.h', 'VT_GETMODE')
-if consio_h
-    features += 'consio.h'
-endif
+features += {'consio.h': consio_h}
 
-fragments = join_paths(source_root, 'waftools', 'fragments')
-
-glibc_thread_name = cc.compiles(files(join_paths(fragments, 'glibc_thread_name.c')),
-                                name: 'glibc-thread-name check')
-if glibc_thread_name and posix
-    features += 'glibc-thread-name'
-endif
+glibc_thread_name = cc.has_function('pthread_setname_np',
+                                    prefix: '#define _GNU_SOURCE\n#include <pthread.h>')
+features += {'glibc-thread-name': glibc_thread_name and posix and not darwin}
 
 osx_thread_name = false
-if not glibc_thread_name
-    osx_thread_name = cc.compiles(files(join_paths(fragments, 'osx_thread_name.c')),
-                                  name: 'osx-thread-name check')
-    if osx_thread_name
-        features += 'osx-thread-name'
-    endif
+if not glibc_thread_name or darwin
+    osx_thread_name = cc.has_function('pthread_setname_np',
+                                      prefix: '#include <pthread.h>')
 endif
+features += {'osx-thread-name': osx_thread_name}
 
 bsd_thread_name = false
 if not osx_thread_name and not glibc_thread_name
-    bsd_thread_name = cc.compiles(files(join_paths(fragments, 'bsd_thread_name.c')),
-                                  name: 'bsd-thread-name check')
-    if bsd_thread_name and posix
-        features += 'bsd-thread-name'
-    endif
-endif
-
-vector = cc.compiles(files(join_paths(fragments, 'vector.c')), name: 'vector check')
-if vector
-    features += 'vector'
-elif get_option('vector').enabled()
-    error('vector enabled but it could not be found!')
+    bsd_thread_name = cc.has_function('pthread_set_name_np',
+                                      prefix: '#include <pthread.h>\n#include <pthread_np.h>')
 endif
+features += {'bsd-thread-name': bsd_thread_name}
 
 bsd_fstatfs = cc.has_function('fstatfs', prefix: '#include <sys/mount.h>\n#include <sys/param.h>')
-if bsd_fstatfs
-    features += 'bsd-fstatfs'
-endif
+features += {'bsd-fstatfs': bsd_fstatfs}
 
 linux_fstatfs = cc.has_function('fstatfs', prefix: '#include <sys/vfs.h>')
-if linux_fstatfs
-    features += 'linux-fstatfs'
-endif
+features += {'linux-fstatfs': linux_fstatfs}
+
+vector = get_option('vector').require(
+    cc.has_type('__attribute__((vector_size(32)))'),
```

Comment:
an attribute is not a type on its own so use `cc.has_type('float __attribute__((vector_size(32)))')`?

Reply:
This is actually slightly awkward, because Meson's `cc.has_function_attribute` internally uses `-Werror=attribute` for GCC/clang, with the note that "XX only warns about unknown or ignored attributes, so force an error".

But per https://mesonbuild.com/Reference-tables.html#function-attributes, vector_size is not something Meson knows how to look up on its own...

Reply:
has_type intrinsically uses this code:

```python
t = f'''{prefix}
    void bar(void) {{
        sizeof({typename});
    }}'''
    return self.compiles(t, env, extra_args=extra_args,
                         dependencies=dependencies)
```

I don't know that that's actually what you want. The function attribute check just checks that actual source code snippets such as this compile:
```c
// visibility:default
int foo(void) __attribute__((visibility("default")));

// fallthrough
int foo( void ) {
  switch (0) {
    case 1: __attribute__((fallthrough));
    case 2: break;
  }
  return 0;
};
```

So you should probably stick with a suitable `cc.compiles()` here, you'll otherwise just get debug logs with `error: expected ')' before '__attribute__'` and "Checking for type "__attribute__((vector_size(32)))" : NO"

Reply:
`float __attribute__((vector_size(32)))` does actually pass (as well as just `attribute((vector_size(32)))`). I'm not sure if that's just a false positive though? But yeah, ideally the function attribute check would have worked but that didn't seem to be supported (always failed when I tried it as well).

Reply:
Ah right, this sizeof stuff compiles with CC but not with CXX.

But also, still, so does this compile successfully:

```
Code:
 
        void bar(void) {
            sizeof(__attribute__((vector_siiize(32))));
        }
Compiler stdout:
 
Compiler stderr:
 /tmp/builddir/meson-private/tmp0wnh4suv/testfile.c: In function 'bar':
/tmp/builddir/meson-private/tmp0wnh4suv/testfile.c:3:13: warning: type defaults to 'int' in type name [-Wimplicit-int]
    3 |             sizeof(__attribute__((vector_siiize(32))));
      |             ^~~~~~
/tmp/builddir/meson-private/tmp0wnh4suv/testfile.c:3:13: warning: 'vector_siiize' attribute directive ignored [-Wattributes]

Checking for type "__attribute__((vector_siiize(32)))" : YES 
```

Reply:
Welp, I guess that answers that. Thanks for the feedback!